### PR TITLE
NFSE-5087 Remove kblock/vblock size params

### PR DESCRIPTION
**Pull request by tristan957**
These are no longer needed and will instead just piggyback on the mblock
size.

Signed-off-by: Tristan Partin <tpartin@micron.com>

## Verified checks?
Have the checks completed?
- [x] meson test -C build --setup=ci
- [x] All commits are signed off

### CODE REVIEW

**Comment by tristan957** [comment]
Not super sure whether the changes in kvset.c are correct. Should I use the kvdb rparam for mblock size there? I am thinking probably, but I will wait for feedback.

**Comment by beckerg** [comment]
I thought we needed to retain configurability to support ZNS drives w/o rebuilding?

**Comment by tristan957** [comment]
I understood the conversation pertaining to mblock size, but maybe my memory is hazy.

**Comment by beckerg** [comment]
Oh, I think we're removing the config params but mpool will set it differently...

**Review by alexttx — CHANGES_REQUESTED** [review]

File: lib/cn/kvset.c [hunk 1]
```diff
@@ -2479,7 +2479,7 @@ kvset_iter_enable_mblock_read_cmn(struct kvset_iterator *iter, struct kblk_reade
     uint  kb_max_sz;
 
     /* compute appropriate node buffer size */
-    kb_max_sz = iter->ks->ks_rp->kblock_size;
+    kb_max_sz = MPOOL_MBLOCK_SIZE_DEFAULT;
```
[the remark this answers is not in the document]
Comment (tristan957) [reply lib/cn/kvset.c:2482]: That is what I thought. Will look into it.

[the remark this answers is not in the document]
Comment (nabeelmmd) [reply lib/cn/kvset.c:2482]: cN can use mpool_mclass_props_get() to query the mblock size for a specific media class. Might be a good idea to store  either the media classes used for storing keys/values or the max kblock/vblock sizes in struct kvset at kvset creation time?

**Review by alexttx — APPROVED** [review]

File: lib/cn/kblock_builder.c [hunk 2]
```diff
@@ -138,10 +138,10 @@ struct curr_kblock {
...
     u32 num_tombstones;
 
-    uint max_size;
-    uint max_pgc;
-    uint blm_pgc;
-    uint wbt_pgc;
+    uint64_t max_size;
+    uint64_t max_pgc;
+    uint     blm_pgc;
+    uint     wbt_pgc;
```
Comment (alexttx) [inline lib/cn/kblock_builder.c:144]: You've switched a subset of these scalars from `uint` to `uint64_t`.  I would switch these back to `uint` or move them all to `uint64_t`.

File: lib/cn/vblock_builder.c [hunk 3]
```diff
@@ -193,10 +188,27 @@ vbb_create(
...
+
+        err = mpool_mclass_props_get(mp, i, &props);
+        if (err) {
+            vlb_free(wbuf, WBUF_LEN_MAX + sizeof(*bld));
+            return err;
+        }
+
+        if (props.mc_mblocksz > bld->max_size)
+            bld->max_size = props.mc_mblocksz;
+    }
```
Comment (alexttx) [inline lib/cn/vblock_builder.c:210]: In theory we can get out of this loop w/ `bld->max_size == 0`.  Maybe change to:
```
for (...) {
    err = mpool_mclass_props_get(mp, i, &props);
    if (err)
        break;
}
if (err || blk->max_size == 0) {
    vlb_free(...);
    return err ? err : merr(UGH);
}
```

File: lib/cn/kblock_builder.c (hunk 2 above)
Comment (tristan957) [reply lib/cn/kblock_builder.c:144]: Well I didn't want to dig too deeply here. kblock_builder is at risk for a lot of integer overflow in my opinion. I only changed what had a dependency on mblocksz although I think these should actually be uint32_t

Comment (beckerg) [reply lib/cn/kblock_builder.c:144]: pgc means page count, so (n * 4096). 256 pages per megabyte, so with kblocks being quite small (32M max???) they should never be in danger of overflowing.  dunno about the others...   does this patch change the max kblock size?

File: lib/cn/kblock_builder.c [hunk 4]
```diff
@@ -154,14 +154,15 @@ struct curr_kblock {
     uint  bloom_used_max;
 };
 
-static HSE_ALWAYS_INLINE uint
+static HSE_ALWAYS_INLINE uint64_t
 free_pgc(struct curr_kblock *kblk)
 {
-    uint used = KBLOCK_HDR_PAGES + HLOG_PGC + kblk->blm_pgc + kblk->wbt_pgc;
+    const uint64_t used = KBLOCK_HDR_PAGES + HLOG_PGC + kblk->blm_pgc + kblk->wbt_pgc;
```
Comment (beckerg) [inline lib/cn/kblock_builder.c:160]: page counts, so should be very small...

File: lib/cn/kblock_builder.c [hunk 5]
```diff
@@ -453,7 +456,7 @@ kblock_init(
     struct kvs_cparams *cp,
     struct kvs_rparams *rp,
     struct perfc_set *  pc,
-    size_t              max_size)
+    uint32_t            max_size)
```
Comment (beckerg) [inline lib/cn/kblock_builder.c:459]: size_t seems right, and is bigger...  why the change?  i hate this tool, no context...

File: lib/cn/kblock_builder.c [hunk 6]
```diff
@@ -1188,9 +1208,9 @@ kbb_finish(struct kblock_builder *bld, struct blk_list *kblks, u64 seqno_min, u6
      */
     if (!kblock_is_empty(&bld->curr)) {
         struct wbb *pt = 0;
-        u64         kbsize = bld->rp->kblock_size;
-        u64         ptsize = (wbb_page_cnt_get(bld->ptree)) * PAGE_SIZE;
-        u64         kbused =
+        uint32_t    kbsize = bld->max_size;
+        uint64_t    ptsize = (wbb_page_cnt_get(bld->ptree)) * PAGE_SIZE;
+        uint64_t    kbused =
```
Comment (beckerg) [inline lib/cn/kblock_builder.c:1213]: Reducing the size of variables runs the risk of quietly truncating the size of a calculation that was previously carried out in 64-bit math to 32-bit math.  Tread very very carefully..

File: lib/cn/spill.c [hunk 7]
```diff
@@ -251,8 +251,6 @@ kv_spill(struct cn_compaction_work *w)
      * readahead buffer).
      */
     direct_read_len = w->cw_rp->cn_compact_vblk_ra;
-    if (direct_read_len < 32 * 1024)
-        direct_read_len = 32 * 1024;
```
Comment (beckerg) [inline lib/cn/spill.c:255]: Why did you eliminate this?

File: lib/cn/kvset.c [hunk 8]
```diff
@@ -2539,15 +2534,12 @@ kvset_iter_enable_mblock_read(struct kvset_iterator *iter)
     struct kblk_reader *kr = &iter->kreader;
     struct vblk_reader *vr;
     void *              mem;
-    uint                vr_buf_sz;
-    uint                vb_max_sz;
+    uint64_t            vr_buf_sz;
     uint                i;
     merr_t              err;
-    int                 ra_size;
+    uint64_t            ra_size;
```
Comment (beckerg) [inline lib/cn/kvset.c:2540]: read-ahead size should never be more than 1048576.  but i haven't looked at this code in a very long time...

File: lib/cn/vblock_builder_internal.h [hunk 9]
```diff
@@ -70,16 +80,16 @@ struct vblock_builder {
...
     void *                     wbuf;
-    uint                       wbuf_off;
-    uint                       wbuf_len;
-    u64                        vgroup;
+    unsigned int               wbuf_off;
+    unsigned int               wbuf_len;
+    uint64_t                   vgroup;
     bool                       destruct;
-    u32                        opt_wrsz;
+    uint64_t                   opt_wrsz;
```
Comment (beckerg) [inline lib/cn/vblock_builder_internal.h:92]: We probably should be using size_t for sizes of things.  I know we don't, but we probably should....

File: lib/cn/kblock_builder.c [hunk 10]
```diff
@@ -1031,17 +1038,30 @@ kbb_create(struct kblock_builder **builder_out, struct cn *cn, struct perfc_set
     bld->pc = pc;
     bld->agegroup = HSE_MPOLICY_AGE_LEAF;
 
+    for (int i = HSE_MCLASS_BASE; i < HSE_MCLASS_COUNT; i++) {
```
Comment (nabeelmmd) [inline lib/cn/kblock_builder.c:1041]: I think we can determine the media class for bld->agegroup + DTYPE_KEY and then query the mblock size from mpool for that media class, like how you did in kblock_finish()?

I don't know what are the consequences of passing the max mblock size of al the configured media classes here...

File: lib/cn/kblock_builder.c [hunk 11]
```diff
@@ -1188,9 +1208,9 @@ kbb_finish(struct kblock_builder *bld, struct blk_list *kblks, u64 seqno_min, u6
      */
     if (!kblock_is_empty(&bld->curr)) {
         struct wbb *pt = 0;
-        u64         kbsize = bld->rp->kblock_size;
-        u64         ptsize = (wbb_page_cnt_get(bld->ptree)) * PAGE_SIZE;
-        u64         kbused =
+        uint32_t    kbsize = bld->max_size;
```
Comment (nabeelmmd) [inline lib/cn/kblock_builder.c:1211]: What's the intent behind using the max mblock size in this function vs determining the right mblock size for the age group of this kvset + DTYPE_KEY?

All kblocks in a kvset are allocated from the same media class as there's no media class allocation fallback support.

File: lib/mpool/src/mpool.c [hunk 12]
```diff
@@ -515,6 +515,12 @@ mpool_mclass_ftw(
     return mclass_ftw(mc, prefix, cb);
 }
 
+bool
+mpool_mclass_is_configured(struct mpool *const mp, const enum hse_mclass mclass)
+{
+    return mpool_mclass_handle(mp, mclass);
+}
+
```
Comment (nabeelmmd) [inline lib/mpool/src/mpool.c:523]: I think "return !!mpool_mclass_handle(mp, mclass)" is clearer...

File: lib/mpool/src/mclass.c [hunk 13]
```diff
@@ -407,7 +407,7 @@ void
 mclass_props_get(struct media_class *const mc, struct mpool_mclass_props *const props)
 {
     props->mc_fmaxsz = mblock_fset_fmaxsz_get(mc->mbfsp);
-    props->mc_mblocksz = mc->mblocksz >> MB_SHIFT;
+    props->mc_mblocksz = mc->mblocksz;
```
Comment (nabeelmmd) [inline lib/mpool/src/mclass.c:410]: Please double check that all the mclass_props_get() call sites does the necessary conversion of props->mc_mblocksz as the unit has changed.

File: lib/cn/vblock_builder_internal.h (hunk 9 above)
Comment (nabeelmmd) [reply lib/cn/vblock_builder_internal.h:92]: And probably off_t for offsets...

File: lib/cn/vblock_builder.c [hunk 14]
```diff
@@ -70,9 +68,7 @@ _vblock_start(struct vblock_builder *bld)
     if (stats)
         count_ops(&stats->ms_vblk_alloc, 1, mbprop.mpr_alloc_cap, get_time_ns() - tstart);
 
-    rp = cn_get_rp(bld->cn);
-
-    if (mbprop.mpr_alloc_cap != rp->vblock_size) {
+    if (mbprop.mpr_alloc_cap != bld->max_size) {
```
Comment (nabeelmmd) [inline lib/cn/vblock_builder.c:71]: I think this check can go away..

File: lib/cn/vblock_builder.c [hunk 15]
```diff
@@ -193,10 +188,27 @@ vbb_create(
     bld->pc = pc;
     bld->ds = cn_get_dataset(cn);
     bld->vgroup = vgroup;
-    bld->max_size = rp->vblock_size;
     bld->agegroup = HSE_MPOLICY_AGE_LEAF;
     bld->wbuf = wbuf;
 
+    mp = cn_get_dataset(cn);
```
Comment (nabeelmmd) [inline lib/cn/vblock_builder.c:194]: Again not sure about the intent for this max logic..

File: lib/cn/spill.c (hunk 7 above)
Comment (tristan957) [reply lib/cn/spill.c:255]: Because I changed the bounds on cn_compact_vblk_ra. If the value can never be less than MPOOL_MBLOCK_SIZE_DEFAULT, then why not use it as ps_min.

Comment (beckerg) [reply lib/cn/spill.c:255]: I'll need to study this to see if it makes sense that direct-read-len is now dependent upon cn_compact_vblk_ra.  Dunno of the top of my head...

File: lib/cn/kblock_builder.c (hunk 10 above)
Comment (tristan957) [reply lib/cn/kblock_builder.c:1041]: For some reason I thought max_size meant the max mblock size. I will adjust.

File: lib/cn/kvset.c (hunk 8 above)
Comment (tristan957) [reply lib/cn/kvset.c:2540]: I am currently just basing it off of cn_compact_{k,v}blk_ra which is a uint64_t. Happy to reduce the type in kvs_rparams.c though.

File: lib/cn/vblock_builder_internal.h (hunk 9 above)
Comment (tristan957) [reply lib/cn/vblock_builder_internal.h:92]: I changed the offsets to off_t. I think we should talk about size_t more broadly since currently we don't expose any size_t parameters. Nabeel and I have talked about this in the past. Happy to change mblocksz to size_t.

File: lib/cn/spill.c (hunk 7 above)
Comment (tristan957) [reply lib/cn/spill.c:255]: Note that the read-ahead values are now the min(mblocksz of all configured media classes) at this point within HSE.

Comment (beckerg) [reply lib/cn/spill.c:255]: Ok, this is coming back to me.  The vblock readahead used to be 256K bytes, which is small because a spill could be reading values from any of several hundred vblocks during any given spill.  The readahead code initiates a readahead as the values in any given buffer are consumed such that half the buffer has been drained (IIRC), so as to try and keep slightly ahead of the reader yet minizmize the memory foot print to perform a spill.  This particular code was trying to deal with values that are larger than the readhead buffer (such large values a typically rare).
So what mblocksz now, typically?

Comment (tristan957) [reply lib/cn/spill.c:255]: HSE currently only supports an mblock size of 32 MB (MPOOL_MBLOCK_SIZE_DEFAULT). There is a check in ikvdb.c.

File: lib/cn/kvset.c (hunk 8 above)
Comment (beckerg) [reply lib/cn/kvset.c:2540]: size_t is probably the right type, does config support size_t?  That said, size_t only for correctness in type, it was never expected be higher than 1048576 in practice.  Still looking at the code...  (int was probably always wrong!)

File: lib/cn/spill.c (hunk 7 above)
Comment (alexttx) [reply lib/cn/spill.c:255]: MPOOL_MBLOCK_SIZE_DEFAULT is 32*1024*1024.  This is 32*1024.  Need to fix `.ps_min` in the param spec.

File: lib/kvs/kvs_rparams.c [hunk 16]
```diff
@@ -269,7 +270,7 @@ static const struct param_spec pspecs[] = {
         },
         .ps_bounds = {
             .as_uscalar = {
-                .ps_min = 0,
+                .ps_min = MPOOL_MBLOCK_SIZE_DEFAULT,
```
Comment (alexttx) [inline lib/kvs/kvs_rparams.c:273]: .ps_min = 32\*1024

Comment (alexttx) [reply lib/kvs/kvs_rparams.c:273]: What happens at runtime if this param is set to an enormous value?

File: lib/cn/spill.c (hunk 7 above)
Comment (tristan957) [reply lib/cn/spill.c:255]: Whoops...32 K

File: lib/cn/kvset.c (hunk 8 above)
Comment (tristan957) [reply lib/cn/kvset.c:2540]: Happy to add support for size_t as a part of this PR

Comment (alexttx) [reply lib/cn/kvset.c:2540]: I'm all for being more standard on scalar types, but let's do that in PRs dedicated to that task.

File: lib/cn/spill.c (hunk 7 above)
Comment (tristan957) [reply lib/cn/spill.c:255]: Then this seems really strange because these read ahead parameters are based on the min(mblocksz of all configured media classes), so it will be in the MB of values regardless because HSE only supports 32MB mblock size. Confused to say the least.

File: lib/kvs/kvs_rparams.c (hunk 16 above)
Comment (beckerg) [reply lib/kvs/kvs_rparams.c:273]: Actually, this should be:
.ps_min = 1024 * 32;
.ps_max = 1024 * 2048;

File: tests/unit/kvs/kvs_rparams_test.c [hunk 17]
```diff
@@ -208,7 +208,7 @@ MTF_DEFINE_UTEST_PRE(kvs_rparams_test, cn_compact_vblk_ra, test_pre)
     ASSERT_EQ((uintptr_t)ps->ps_stringify, (uintptr_t)param_default_stringify);
     ASSERT_EQ((uintptr_t)ps->ps_jsonify, (uintptr_t)param_default_jsonify);
     ASSERT_EQ(256 * 1024, params.cn_compact_vblk_ra);
-    ASSERT_EQ(0, ps->ps_bounds.as_uscalar.ps_min);
+    ASSERT_EQ(MPOOL_MBLOCK_SIZE_DEFAULT, ps->ps_bounds.as_uscalar.ps_min);
```
Comment (beckerg) [inline tests/unit/kvs/kvs_rparams_test.c:211]: I think we need checks here to ensure the default value is within bounds, because in this case I think it would have failed...

File: lib/cn/kvset.c [hunk 18]
```diff
@@ -2539,15 +2534,12 @@ kvset_iter_enable_mblock_read(struct kvset_iterator *iter)
...
 
     ra_size = iter->ks->ks_rp->cn_compact_vblk_ra;
-    if (ra_size < 32 * 1024)
-        ra_size = 32 * 1024;
```
Comment (beckerg) [inline lib/cn/kvset.c:2550]: This is fine as long as  cn_compact_vblk_ra.ps_min == 32 * 1024;

File: lib/cn/kvset.c [hunk 19]
```diff
@@ -2557,21 +2549,17 @@ kvset_iter_enable_mblock_read(struct kvset_iterator *iter)
...
     /* Limit buffered reads to values lesser than cn_compact_vblk_ra. The
      * upper level spill/compaction routines make direct reads for the sizes
      * matching or exceeding cn_compact_vblk_ra.
      */
-    vr_buf_sz = max_t(uint, PAGE_SIZE, ra_size);
-
-    vb_max_sz = iter->ks->ks_rp->vblock_size;
-    if (vb_max_sz > VLB_ALLOCSZ_MAX / 2)
-        vb_max_sz = VLB_ALLOCSZ_MAX / 2;
+    vr_buf_sz = max_t(uint64_t, PAGE_SIZE, ra_size);
```
Comment (beckerg) [inline lib/cn/kvset.c:2559]: Since ra_size should never be less than 32K it seems this line could just be:   vr_buf_sz = ra_size;

File: lib/cn/spill.c (hunk 7 above)
Comment (beckerg) [reply lib/cn/spill.c:255]: The readahead values shouldn't be at all linked to mblock sizes...  Where do you see that?  Per my .ps_max comment, the vblock readahead buffer should never be larger than 2M (but really 2x4M because there are actually two buffers within one allocation, one being looked at by spill and the other undergoing read, where they swap when the spill-read buffer is exhausted).  But .ps_max should set the max to 2M and this code will double it.

**Review by beckerg — APPROVED** [review]

**Comment by beckerg** [comment]
Wait, I need to look at that direct-read change again...

File: lib/mpool/src/mclass.c (hunk 13 above)
Comment (beckerg) [reply lib/mpool/src/mclass.c:410]: Why change it?  I don't think we ever want mblocks to be a fraction of 1M...

Comment (tristan957) [reply lib/mpool/src/mclass.c:410]: I changed it because mpool receives the mblocksz from the params in bytes, so it made no sense to return mblocksz in MB.

File: lib/kvs/kvs_rparams.c [hunk 20]
```diff
@@ -307,12 +308,12 @@ static const struct param_spec pspecs[] = {
...
         .ps_default_value = {
-            .as_uscalar = 512 * 1024,
+            .as_uscalar = 512 << KB_SHIFT,
         },
         .ps_bounds = {
             .as_uscalar = {
-                .ps_min = 0,
-                .ps_max = UINT64_MAX,
+                .ps_min = 32 << KB_SHIFT,
+                .ps_max = 2 << MB_SHIFT,
```
Comment (beckerg) [inline lib/kvs/kvs_rparams.c:316]: I would stick with just one shift unit (KB_SHIFT) as it makes size comparison to the reader a bit simpler.  But this is fine..

File: lib/mpool/src/mclass.c (hunk 13 above)
Comment (beckerg) [reply lib/mpool/src/mclass.c:410]: Probably specifying mblocksz in units of bytes is either wrong or not very useful (or both!), but we can address that some other time...

File: scripts/build/utpp [hunk 21]
```diff
@@ -191,7 +191,7 @@ function emit_header(name, set_name, ku, prefix, args)
             printf("\t\trc = (void *)(mtfm_%s_%s_get())(__VA_ARGS__); \\\n",
                    set_name, name) > TARGET_FILE;
             printf("\trc; \\\n") > TARGET_FILE;
-        } else if (prefix ~ /bool|int|uint|unsigned|long|ulong|merr_t|size_t|ssize_t|enum|s8|s16|s32|s64|u8|u16|u32|u64|int8_t|int16_t|int32_t|int64_t|uint8_t|uint16_t|uint32_t|uint64_t/) {
+        } else if (prefix ~ /bool|int|uint|unsigned|unsigned int|long|ulong|merr_t|size_t|ssize_t|enum|s8|s16|s32|s64|u8|u16|u32|u64|int8_t|int16_t|int32_t|int64_t|uint8_t|uint16_t|uint32_t|uint64_t/) {
```
Comment (beckerg) [inline scripts/build/utpp:194]: Add "unsigned long" while you're here?

File: lib/cn/vblock_builder.c [hunk 22]
```diff
@@ -310,10 +310,25 @@ vbb_finish(struct vblock_builder *bld, struct blk_list *vblks)
...
+    policy = cn_get_mclass_policy(bld->cn);
+
+    err = mpool_mclass_props_get(
+        bld->ds, policy->mc_table[bld->agegroup][HSE_MPOLICY_DTYPE_KEY], &props);
+    if (err)
+        return err;
+
+    bld->max_size = props.mc_mblocksz;
+
+    return err;
```
Comment (beckerg) [inline lib/cn/vblock_builder.c:331]: I think I see a good many calls in spill.c that do not check this new return code (via kvset_builder_set_agegroup).  Maybe I missed something?
What does it mean if you set the agegroup but the props-get call fails?  The agegroup has been modified, but the max_size has not..

Comment (alexttx) [reply lib/cn/vblock_builder.c:331]: Also, that should not be DTYPE_KEY.

Comment (alexttx) [reply lib/cn/vblock_builder.c:331]: Maybe @nabeelmmd can fix the DTYPE in his PR

File: lib/kvs/kvs_rparams.c (hunk 20 above)
Comment (tristan957) [reply lib/kvs/kvs_rparams.c:316]: I will keep this in mind for a future cleanup. I can see your point

File: lib/cn/vblock_builder.c (hunk 22 above)
Comment (nabeelmmd) [reply lib/cn/vblock_builder.c:331]: Yes, will fix it in my PR.
